src/dmd/backend/cod3.d

```diff
@@ -6812,28 +6811,88 @@ nothrow:
     {
         // Emit accumulated bytes to code segment
         debug assert(index < bytes.length);
-        offset += objmod.bytes(seg, offset, cast(uint)index, bytes.ptr);
+
+        if (disasmBuf)                     // write to buffer for disassembly
+        {
+            foreach (c; bytes[0 .. index]) // not efficient, but for verbose output anyway
+                disasmBuf.push(c);
+        }
+
+        offset += objmod.bytes(seg, offset, index, bytes.ptr);
         index = 0;
     }
 
     @trusted
-    void gen(char c) { bytes[index++] = c; }
+    void gen(ubyte c) { bytes[index++] = c; }
 
     @trusted
-    void genp(size_t n, void *p) { memcpy(&bytes[index], p, n); index += n; }
+    void genp(uint n, void *p) { memcpy(&bytes[index], p, n); index += n; }
 
     @trusted
     void flush() { if (index) flushx(); }
 
     @trusted
-    uint getOffset() { return cast(uint)(offset + index); }
+    uint getOffset() { return offset + index; }
 
     @trusted
-    uint available() { return cast(uint)(bytes.sizeof - index); }
+    uint available() { return cast(uint)bytes.length - index; }
+
+    /******************************
+     * write64/write32/write16 write `value` to `disasmBuf`
+     */
+    @trusted
+    void write64(ulong value)
+    {
+        if (disasmBuf)
+        {
+            disasmBuf.push(cast(ubyte)value);
+            disasmBuf.push(cast(ubyte)(value >>  8));
+            disasmBuf.push(cast(ubyte)(value >> 16));
+            disasmBuf.push(cast(ubyte)(value >> 24));
+            disasmBuf.push(cast(ubyte)(value >> 32));
+            disasmBuf.push(cast(ubyte)(value >> 36));
+            disasmBuf.push(cast(ubyte)(value >> 40));
+            disasmBuf.push(cast(ubyte)(value >> 44));
+        }
+    }
+
+    pragma(inline, true)
+    @trusted
+    void write32(uint value)
+    {
+        if (disasmBuf)
+        {
+            disasmBuf.push(cast(ubyte)value);
+            disasmBuf.push(cast(ubyte)(value >>  8));
+            disasmBuf.push(cast(ubyte)(value >> 16));
+            disasmBuf.push(cast(ubyte)(value >> 24));
+        }
+    }
+
+    pragma(inline, true)
+    @trusted
+    void write16(uint value)
+    {
+        if (disasmBuf)
+        {
+            disasmBuf.push(cast(ubyte)value);
+            disasmBuf.push(cast(ubyte)(value >> 8));
+        }
+    }
 }
 
+/**************************
+ * Convert instructions to object code and write them to objmod.
+ * Params:
+ *      seg = code segment to write to, code starts at Offset(seg)
+ *      c = list of instructions to write
+ *      disasmBuf = if not null, then also write object code here
```

Comment:
The looks fragile, the function now outputs in two places which can go out of sync. Imagine debugging code generation with this feature and finding out the `-vasm` assembly output is not actually what the compiler generated.

Reply:
It was the easiest and least intrusive way to do it, by far.

Reply:
It's not like you're outputting backend IR, disassembly is not something DMD specific, I don't see why it needs to be interwoven with the backend.

Reply:
> It's not like you're outputting backend IR, disassembly is not something DMD specific, I don't see why it needs to be interwoven with the backend.

GCC and LLVM both already do this in such a way that I can't see anyone using this if it were not specific to dmd, especially if consider they'd be adding a fairly large blob of code just to support X86 (but not the latest instructions) but not any of their other targets.

Reply:
disasm86 supports the latest instructions, just like the inline assembler does. I agree this is a dmd backend specific feature, not a D feature, and what ldc and gdc decide to do about it is entirely up to them.

Reply:
> What if you disassemble the .obj file around the time it would be written to disk?

Because of the complexity of the OMF file format, again, this way is the easiest and least intrusive way to do it, by far. I also don't have to write a separate decoder for each supported object file format.